pyvisa_py/protocols/usbtmc.py

```diff
@@ -479,21 +480,29 @@ def read(self, size):
             try:
                 resp = raw_read(recv_chunk + header_size + max_padding)
                 response = BulkInMessage.from_bytes(resp)
-                received.extend(response.data)
-                while len(resp) == self.usb_recv_ep.wMaxPacketSize:
+                received_transfer.extend(response.data)
+                while (
+                    len(resp) == self.usb_recv_ep.wMaxPacketSize
+                    or len(received_transfer) < response.transfer_size
+                ):
                     # USBTMC Section 3.3 specifies that the first usb packet
                     # must contain the header. the remaining packets do not need
                     # the header the message is finished when a "short packet"
                     # is sent (one whose length is less than wMaxPacketSize)
+                    # wMaxPacketSize may be incorrectly reported by certain drivers.
+                    # Therefore, continue reading until the transfer_size is reached.
                     resp = raw_read(recv_chunk + header_size + max_padding)
-                    received.extend(resp)
+                    received_transfer.extend(resp)
+
+                    # Detect EOM only when device sends all expected bytes.
+                if len(received_transfer) >= response.transfer_size:
+                    eom = response.transfer_attributes & 1
+                    # Truncate data to the specified length (discard padding)
+                    # USBTMC header (12 bytes) has already truncated
+                    received_message = received_transfer[: response.transfer_size]
```

Comment:
```suggestion
                    received_message.extend(received_transfer[: response.transfer_size])
```

Reply:
you are correct the data needs to be added.